utils: add disk_space_monitor

bhalevy opened the pull request:
The purpose of this PR is to implement fundamental infrastructure for disk-space monitoring,
onto which services can subscribe and act upon the collected information, like increasing compaction shares and reducing min_threshold to 2 when capacity is low, or increasing streaming shares to facilitate faster bootstrap for scaling out when capacity runs low.  Also, guardrail can be implemented to reject writes above e.g. 98% utilization and allow only delete and truncate.  snapshots can deleted as well when capacity runs low. Etc.

Instantiated only on shard 0.
And add unit test.

Currently, only subscribe from unit test

Ref https://github.com/scylladb/scylladb/issues/3524

* New feature, no backport needed

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13158/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13158/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/214/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13158/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4755/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1678/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12779/)
#### Failed Tests (11974/12934):
- [test_auth_no_quorum.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.debug.1)
- [test_auth_no_quorum.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.debug.2)
- [test_auth_no_quorum.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.debug.3)
- [test_auth_no_quorum.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.dev.1)
- [test_auth_no_quorum.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.dev.2)
- [test_auth_no_quorum.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.dev.3)
- [test_auth_no_quorum.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.release.1)
- [test_auth_no_quorum.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.release.2)
- [test_auth_no_quorum.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.release.3)
- [test_auth_raft_snapshot_transfer.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13158/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer.debug.1)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In [31ec529](https://github.com/scylladb/scylladb/commit/31ec52998263c74ecb0aef5e738c06ff4cee7968):
- Fixed use-after-move in main
- Aggregate stats over std::set<fs::path>

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13164/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13164/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/218/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1685/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12784/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4761/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13164/)
#### Build Details:
- Duration: 3 hr 42 min
- Builder: spider4.cloudius-systems.com

bhalevy commented:
[6100992](https://github.com/scylladb/scylladb/commit/610099202955216a0a61dc7cad07b1d710ec3575):
- added new source file to utils/CMakeLists.txt

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13194/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13194/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/246/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4790/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12812/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1715/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13194/)
#### Build Details:
- Duration: 3 hr 43 min
- Builder: spider2.cloudius-systems.com

paszkow commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,83 @@
+    if (du >= _next_above || du < _next_below) {
+        logger.info("total_space={} free_space={} utilization={:.1f}%", dsm.total_space(), dsm.free_space(), du * 100);
+        if (du >= 0) {
+            auto free_du = std::min(1. - du, 0.);
```
`du` is a value between 0 and 1 or -1 if total disk space is unset. Then this value is always 0.

bhalevy replied:
good catch.  thanks

bhalevy replied:
Fixed in [8616db6](https://github.com/scylladb/scylladb/commit/8616db640cbca07d42990522d95f3b2983c3197c)

paszkow commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,83 @@
+    subscription_list_type::iterator next;
+    for (auto it = _subscriptions.begin(); it != _subscriptions.end(); it = next) {
+        next = it;
+        ++next;
```
What's the point of `next` here? Can't we simply do `++it`?

bhalevy replied:
I wanted to allow the callback to destroy the subscription, which would invalidate the iterator.l, hence next is calculated before calling the subscribed function.

bhalevy replied:
Explained in comment in [8616db6](https://github.com/scylladb/scylladb/commit/8616db640cbca07d42990522d95f3b2983c3197c)

paszkow commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,127 @@
+        scheduling_group sched_group;
+        lowres_clock::duration normal_polling_interval = 10s;
+        lowres_clock::duration high_polling_interval = 1s;
+        // Use high_polling_interval above this thrshold
```
Typo: `threshold`

bhalevy replied:
Fixed in [8616db6](https://github.com/scylladb/scylladb/commit/8616db640cbca07d42990522d95f3b2983c3197c)

bhalevy commented:
In v4 ([8616db6](https://github.com/scylladb/scylladb/commit/8616db640cbca07d42990522d95f3b2983c3197c)):
- fixed std::min(1.-du, 0.) to std::max and added comment about why it is needed
- added comment about why next iterator is needed before invoking the callback
- fixed type: s/thrshold/threshold/

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13262/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13262/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/297/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1789/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4879/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12893/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13262/)
#### Build Details:
- Duration: 3 hr 54 min
- Builder: spider3.cloudius-systems.com

raphaelsc commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,127 @@
+        lowres_clock::duration high_polling_interval = 1s;
+        // Use high_polling_interval above this threshold
+        float polling_interval_threshold = 0.95;
+    };
```
The file system can start behaving very very bad at 95%, failing allocations, etc. We're very sensitive to latency and latency can increase from the XFS side. Shouldn't we treat 90% as the upper bound? Otherwise, we're saying that we treat range (0.9, 0.95) as acceptable.

raphaelsc replied:
If we're pushing towards 90% density, it means when we cross it, balancer will start migrating tablets away from the node. So the disk monitor, which is a replica entity, should treat 80-85% usage as high, so it can do local things to avoid hitting the balancer threshold.

bhalevy replied:
we can lower this interval, though migration should kick-in earlier than that (and it happens externally, using the node-exporter, not via this `disk_space_monitor`)

raphaelsc replied:
yes, I was more worried about this monitor not having an actual effect, since balancer might kick in earlier. btw, we already call each node periodically to pull per-table stats for split / merge decisions, the interface can be extended too include per-node stats in additional to per-table ones I think.

raphaelsc replied:
see storage_service::load_stats_for_tablet_based_tables(), this is the handler

avikivity replied:
These should all be config items

bhalevy replied:
> we already call each node periodically to pull per-table stats for split / merge decisions, the interface can be extended too include per-node stats in additional to per-table ones I think.

Let's wait with that until we do something with this information.
I'm not sure that extending load_stats_for_tablet_based_tables makes sense since it provides per-table stats.
I think that a new rpc would be more appropriate.

raphaelsc replied:
> > we already call each node periodically to pull per-table stats for split / merge decisions, the interface can be extended too include per-node stats in additional to per-table ones I think.
> 
> Let's wait with that until we do something with this information. I'm not sure that extending load_stats_for_tablet_based_tables makes sense since it provides per-table stats. I think that a new rpc would be more appropriate.

```
struct load_stats {
    std::unordered_map<table_id, table_load_stats> tables;

    load_stats& operator+=(const load_stats& s);
    friend load_stats operator+(load_stats a, const load_stats& b) {
        return a += b;
    }
};
```

we feed load_stats into the balancer (pulled periodically by the coordinator), so it makes sense to extend load_stats with map<host_id, node_load_stats>. but anyway, we don't have to make a decision now.

raphaelsc replied:
we could reuse it in the future for cpu balancer, etc.

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,127 @@
+    };
+
+    using subscription_callback_type = noncopyable_function<future<> (const disk_space_monitor&)>;
+    class subscription : public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
```
What about boost::signals2? Aren't you duplicating it here?

bhalevy replied:
Hmm, okay.  Can do.  I wasn't aware of it, but I see we already use it e.g. to get snitch updates

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,127 @@
+    };
+
+private:
+    std::set<std::filesystem::path> _data_dirs;
```
vector?

bhalevy replied:
Sure, no problem.

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,127 @@
+    uint64_t total_space() const noexcept { return _total_space; }
+    uint64_t free_space() const noexcept { return _free_space; }
+    float disk_utilization() const noexcept {
+        return _total_space ? (float)(_total_space - _free_space) / _total_space : -1;
```
This isn't meaningful, each mount has to be managed separately.

Your average utilization can be 80% and one of the mounts at 100%.

bhalevy replied:
okay, though it's all theoretical.
We currently support only a single dir in pratice.
Maybe for simplicity reasons we can start with a single dir, and extend it to multiple directories if and when we really support that end-to-end.

avikivity replied:
Let's not add more breakage.

If this thing supports multiple directories, it should support them correctly.

bhalevy replied:
We don't have to support multiple directories at the moment.

raphaelsc replied:
another possible usage of disk monitor is object storage. we monitor fast storage, and if filling up, we throttle incoming writes in order to allow data to be demoted to slow storage without ENOSPC. /cc @tchaikov

avikivity commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,84 @@
+
+future<> disk_space_monitor_logger::log(const disk_space_monitor& dsm) {
+    auto du = dsm.disk_utilization();
+    if (du >= _next_above || du < _next_below) {
```
Logging should be left to the caller who can provide context.

bhalevy replied:
This is just an exemplary caller to periodically log changes in disk utilization, even when it's not critical.

avikivity replied:
I don't understand what an exemplary caller is.

bhalevy replied:
It's `disk_space_monitor_logger`.
I'll just drop it in the next version as it's not important at this stage.

avikivity commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,84 @@
+}
+
+future<> disk_space_monitor::poll() {
+    auto gh = _async_gate.hold();
```
It's better to run the entire loop in a coroutine and wait for it in stop(). Overall it tends to be simpler.

bhalevy replied:
Sounds good, thanks.

avikivity commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,84 @@
+    _data_dirs = std::move(data_dirs);
+    _cfg = std::move(cfg);
+    _timer.set_callback(cfg.sched_group, [this] { (void)poll(); });
+    (void)poll();
```
Wait for this at the end.

avikivity commented on test/boost/disk_space_monitor_test.cc:
```diff
@@ -0,0 +1,40 @@
+
+    while (!called) {
+        yield().get();
+    }
```
What is being tested here?

raphaelsc replied:
I wonder if this test ever finishes without external action. We could maybe set up a small fuse file system (similar to how we set up a env for s3 tests) to this test, and the test can allocate until monitor calls back the subscriber that will deallocate space, then we can check that the monitor will not call back in the next interval

bhalevy replied:
> What is being tested here?

That the disk_space_monitor setup and teardown work correclty, plus the registered callback is being called.

bhalevy replied:
> I wonder if this test ever finishes without external action. We could maybe set up a small fuse file system (similar to how we set up a env for s3 tests) to this test, and the test can allocate until monitor calls back the subscriber that will deallocate space, then we can check that the monitor will not call back in the next interval

The callbacks are called periodically and unconditionally.
Why complicate the test?

raphaelsc replied:
> > I wonder if this test ever finishes without external action. We could maybe set up a small fuse file system (similar to how we set up a env for s3 tests) to this test, and the test can allocate until monitor calls back the subscriber that will deallocate space, then we can check that the monitor will not call back in the next interval
> 
> The callbacks are called periodically and unconditionally.
> Why complicate the test?

Ah usually this kind of interface only calls its subscribers when some event happens. Not saying it's necessarily a bad thing just that I was surprised

bhalevy replied:
That was my initial intent but eventually I came to conclusion that inplementing a  generic threshold based event system would be too complicated and it would be more straightforward for each subscriber to manage its own state, if needed, or just implement a stateless algorithm, for example, the maintenance shares controller, or the write guardrail (say to issue warnings above a soft threshold and reject writes above the hard threshold)

bhalevy replied:
Next version tests space monitoring (via simulation, not of a real filesystem)

bhalevy commented:
In v5 ([e163a5a](https://github.com/scylladb/scylladb/commit/e163a5adf7a6177f6ef6d10b88ab53b541a66747)):
- rebased
- templatized and vitualized the class for testing
  - improved unit test
- used boost::signals2 for notifications
- added configuration items to db/config
- dropped the disk_space_logger

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13379/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13379/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
#### Failed Tests (1/1):
- [test_disk_space_monitor](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13379/testReport/junit/boost/disk_space_monitor_test/test_disk_space_monitor) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disk_space_monitor)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: i-0a8dc52569e9327f2 (m5ad.8xlarge)

avikivity commented:
> added configuration items to db/config

They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.

Utilities should be configured via a struct.

bhalevy commented:
> > added configuration items to db/config
> 
> They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.
> 
> Utilities should be configured via a struct.

I'm not I understand what you aim at.
Do you mean that the disk_space_monitor should keep updateable_value:s so that they can be updated at runtime?

avikivity commented:
> > > added configuration items to db/config
> > 
> > 
> > They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.
> > Utilities should be configured via a struct.
> 
> I'm not I understand what you aim at. Do you mean that the disk_space_monitor should keep updateable_value:s so that they can be updated at runtime?

If it's a db::config item, the thing is bound to exactly one user. You can't instantiate two because they'll have the same configuration.

It needs to expose a config struct, which its user can populate from db::config. Two users can populate their configs from different config items.

avikivity commented:
> > > > > added configuration items to db/config
> > > > 
> > > > 
> > > > They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.
> > > > Utilities should be configured via a struct.
> > > 
> > > 
> > > I'm not I understand what you aim at. Do you mean that the disk_space_monitor should keep updateable_value:s so that they can be updated at runtime?
> > 
> > 
> > If it's a db::config item, the thing is bound to exactly one user. You can't instantiate two because they'll have the same configuration.
> > It needs to expose a config struct, which its user can populate from db::config. Two users can populate their configs from different config items.
> 
> Can you please give an example of what you see as a user? Why multiple users would need independent configurations of the polling rate and its threshold? 

If you put something in utils/, you have to be prepared to answer such questions yourself. Why are you introducing general utilities if you can't imagine a use case for them?



> I understand a possible need to make the callback logic a bit more sophisticated, like having a tolerancd so that the callback function would be invoked when the disk free space ratio changes by more than the tolerance, but I'm not sure what a complex control over polling would buy us.

avikivity commented:
> > > > > > added configuration items to db/config
> > > > > 
> > > > > 
> > > > > They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.
> > > > > Utilities should be configured via a struct.
> > > > 
> > > > 
> > > > I'm not I understand what you aim at. Do you mean that the disk_space_monitor should keep updateable_value:s so that they can be updated at runtime?
> > > 
> > > 
> > > If it's a db::config item, the thing is bound to exactly one user. You can't instantiate two because they'll have the same configuration.
> > > It needs to expose a config struct, which its user can populate from db::config. Two users can populate their configs from different config items.
> > 
> > 
> > Can you please give an example of what you see as a user? Why multiple users would need independent configurations of the polling rate and its threshold?
> 
> If you put something in utils/, you have to be prepared to answer such questions yourself. Why are you introducing general utilities if you can't imagine a use case for them?
> 


Imagine we modularized the system. Now modularization is broken because utils/disk_space_monitor depends on db/config and db/config depends on utils/config_file

bhalevy commented:
> > > > > > > added configuration items to db/config
> > > > > > 
> > > > > > 
> > > > > > They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.
> > > > > > Utilities should be configured via a struct.
> > > > > 
> > > > > 
> > > > > I'm not I understand what you aim at. Do you mean that the disk_space_monitor should keep updateable_value:s so that they can be updated at runtime?
> > > > 
> > > > 
> > > > If it's a db::config item, the thing is bound to exactly one user. You can't instantiate two because they'll have the same configuration.
> > > > It needs to expose a config struct, which its user can populate from db::config. Two users can populate their configs from different config items.
> > > 
> > > 
> > > Can you please give an example of what you see as a user? Why multiple users would need independent configurations of the polling rate and its threshold?
> > 
> > 
> > If you put something in utils/, you have to be prepared to answer such questions yourself. Why are you introducing general utilities if you can't imagine a use case for them?

The possible use cases are described in the PR description.  My plan is to have a single disk space monitor instance, started at main() that will have multiple subscribers, not an instance per subscriber.

> Imagine we modularized the system. Now modularization is broken because utils/disk_space_monitor depends on db/config and db/config depends on utils/config_file

But utils/disk_space_monitor doesn't depend on db/config. It's `config` structure contains trivial members in this version, and only its user - main.cc depends on db/config to get the values for it.

Even if we pass updateable_value members in d_s_m::config, it would depend on utils/*, not on db/config

raphaelsc commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,131 @@
+
+    void start(std::filesystem::path data_dir) {
+        _data_dir = std::move(data_dir);
+        _poller_fut = poll();
```
nit: we could have passed data dir in the configuration.

bhalevy replied:
yes, or pass it to the constructor

bhalevy commented:
In v6 ([9d4d053](https://github.com/scylladb/scylladb/commit/9d4d053d0bc43a114df156e891e664d64d81a641)):
- used utils::updateable_value for disk_space_monitor::config members
  - set to const value by unit test
  - refers to named_value in db::config in main/cql_test_env
- pass data_dir to constructor (as standalone parameter, not as config member)

avikivity commented:
> > > > > > > > added configuration items to db/config
> > > > > > > 
> > > > > > > 
> > > > > > > They should be config items at the point of call, not here. If something refers to a config item, it's not a utility.
> > > > > > > Utilities should be configured via a struct.
> > > > > > 
> > > > > > 
> > > > > > I'm not I understand what you aim at. Do you mean that the disk_space_monitor should keep updateable_value:s so that they can be updated at runtime?
> > > > > 
> > > > > 
> > > > > If it's a db::config item, the thing is bound to exactly one user. You can't instantiate two because they'll have the same configuration.
> > > > > It needs to expose a config struct, which its user can populate from db::config. Two users can populate their configs from different config items.
> > > > 
> > > > 
> > > > Can you please give an example of what you see as a user? Why multiple users would need independent configurations of the polling rate and its threshold?
> > > 
> > > 
> > > If you put something in utils/, you have to be prepared to answer such questions yourself. Why are you introducing general utilities if you can't imagine a use case for them?
> 
> The possible use cases are described in the PR description. My plan is to have a single disk space monitor instance, started at main() that will have multiple subscribers, not an instance per subscriber.
> 
> > Imagine we modularized the system. Now modularization is broken because utils/disk_space_monitor depends on db/config and db/config depends on utils/config_file
> 
> But utils/disk_space_monitor doesn't depend on db/config. It's `config` structure contains trivial members in this version, and only its user - main.cc depends on db/config to get the values for it.
> 
> Even if we pass updateable_value members in d_s_m::config, it would depend on utils/*, not on db/config


I was too hasty - I reviewed the changelog and not the patch, I forgot the patch wasn't just the new utility, but also included the integration into main.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13469/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13469/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
#### Failed Tests (1/1):
- [test_disk_space_monitor](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13469/testReport/junit/boost/disk_space_monitor_test/test_disk_space_monitor) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disk_space_monitor)
#### Build Details:
- Duration: 41 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
In v7 ([8c3f36f](https://github.com/scylladb/scylladb/commit/8c3f36f34cea60962aa9dac2f33c4bbd120b90eb))
- rebased
- more accurate sleep in poller (correct sleep time based on elapsed time between loops)
- adjusted unit test to work in debug mode by advancing manual clock in smaller increments

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13507/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13507/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/446/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5075/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13091/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1967/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13507/)
#### Build Details:
- Duration: 3 hr 59 min
- Builder: spider8.cloudius-systems.com

tchaikov commented on main.cc:
```diff
@@ -1151,6 +1153,28 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+                .polling_interval_threshold = cfg->disk_space_monitor_polling_interval_threshold,
+            };
+            fs::path data_dir;
+            if (data_dir_set.get_paths().size() == 0) {
```
nit, i'd suggest just put:
```c++
if (data_dir_set.empty()) {
    throw ...;
}
fs::path data_dir = *data_dir_set.get_paths().begin();
```

because

- if we don't care about the exact number of elements in the return value `get_paths()`, we can just check for its emptiness
- less indent level, as the branch with `throw` is practically an early return.
- smaller lexical scope of `data_dir` for better readability.

bhalevy replied:
done in v8

tchaikov commented on main.cc:
```diff
@@ -727,6 +728,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
     sharded<locator::effective_replication_map_factory> erm_factory;
     sharded<service::migration_notifier> mm_notifier;
     sharded<service::endpoint_lifecycle_notifier> lifecycle_notifier;
+    std::optional<utils::disk_space_monitor<lowres_clock>> dsm;
```
could you please consider using a more descriptive function name like `disk_space_monitor` instead of the current abbreviated name. While the codebase currently uses abbreviations such as `cm`, `ss`, and `mm` in `main.cc`, a more comprehensive name would enhance code readability, especially for a complex function spanning over 1000 lines.

bhalevy replied:
changed to `disk_space_monitor` in v8

bhalevy commented:
In v8 ([63a4487](https://github.com/scylladb/scylladb/commit/63a4487aad78078a725fbf901d78ccd907fe481a)):
- rebased
- fixed review comments from @tchaikov 
- changed signature of directories::set::get_paths to `const std::set<fs::path>& get_paths() const noexcept`
  - as there's no reason to copy the set (especially for .empty() or .size())
- fixed logger definition to `dsmlog("disk_space_monitor")`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13670/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13670/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/556/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5219/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2111/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13235/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13670/)
#### Failed Tests (1/36085):
- [test_index_doesnt_flood_cache_in_small_partition_workload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13670/testReport/junit/boost/cache_algorithm_test/test_index_doesnt_flood_cache_in_small_partition_workload) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_doesnt_flood_cache_in_small_partition_workload)
#### Build Details:
- Duration: 8 hr 6 min
- Builder: i-0e7ede142c3599d00 (m5ad.8xlarge)

avikivity commented on main.cc:
```diff
@@ -727,6 +728,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
     sharded<locator::effective_replication_map_factory> erm_factory;
     sharded<service::migration_notifier> mm_notifier;
     sharded<service::endpoint_lifecycle_notifier> lifecycle_notifier;
+    std::optional<utils::disk_space_monitor<lowres_clock>> disk_space_monitor;
```
Why optional?

avikivity replied:
Please add _shard0 so it's clear it's only usable there.

bhalevy replied:
To contro its construction, similar to sharded services.

bhalevy replied:
> Please add _shard0 so it's clear it's only usable there.

Done in v9 ([733ca78](https://github.com/scylladb/scylladb/commit/733ca78ff5d46303b1250c9c595f7668083a239d))

avikivity commented on test/boost/disk_space_monitor_test.cc:
```diff
@@ -0,0 +1,132 @@
+
+    BOOST_REQUIRE_EQUAL(wait_for_call(), 4);
+    require_in_range(calls[3].t, calls[2].t, dsm_config.high_polling_interval);
+    BOOST_REQUIRE_EQUAL(calls[3].util, dsm_config.polling_interval_threshold);
```
What are we testing here?

bhalevy replied:
The monitor's polling interval as a function of (simulated) disk utilization.

Not the measurement of thereof.

avikivity replied:
Well, we should test the actual measurement.

Otherwise, "return 0" would pass the tests too.

bhalevy replied:
The problem I saw with this is that if we use a live filesystem, e.g. /tmp then its utilization might change while we measure it, and if we want a clean filesystem, say using a loopback mount, we'll need to be sudoers to mount this filesystem for the test.  If that's ok, then it's doable.

bhalevy replied:
Edit: I found `udisksctl` that seems to allow setting up and mounting a loop device in user mode.
Will send a version using it.
It should be added to the frozen toolchain

bhalevy replied:
Cool! Thanks for pointing that out.
I'll try to follow its steps then.

bhalevy replied:
Done in v10 ([9a3db9a](https://github.com/scylladb/scylladb/commit/9a3db9ac9e444e6ef85f52f0223d11cf6d89b4e2))

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,137 @@
+
+using namespace std::chrono_literals;
+namespace bs2 = boost::signals2;
+
```
Do not add short names in headers. They spread everywhere.

bhalevy replied:
ok

raphaelsc replied:
I also advise against including namespace in header since it can cause conflict. It happened to me :)

bhalevy replied:
> Do not add short names in headers. They spread everywhere.

Dropped in v9 ([733ca78](https://github.com/scylladb/scylladb/commit/733ca78ff5d46303b1250c9c595f7668083a239d))

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13670/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13670/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/disk_space_monitor_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/556/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5219/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2111/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13235/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13670/)
> 
> #### Failed Tests (1/36085):
> * [test_index_doesnt_flood_cache_in_small_partition_workload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13670/testReport/junit/boost/cache_algorithm_test/test_index_doesnt_flood_cache_in_small_partition_workload) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_doesnt_flood_cache_in_small_partition_workload)
> 

Unrelated: https://github.com/scylladb/scylladb/issues/21536

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,137 @@
+extern seastar::logger dsmlog;
+
+// Instantiated only on shard 0
+template <typename Clock = lowres_clock>
```
Why a template? I see you use manual_clock in the test, but I don't understand what you're testing.

bhalevy replied:
The test is the only reason for the template indeed.

bhalevy replied:
Template dropped in v10 ([9a3db9a](https://github.com/scylladb/scylladb/commit/9a3db9ac9e444e6ef85f52f0223d11cf6d89b4e2))

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,137 @@
+
+private:
+    future<> poll() {
+        return async([this] {
```
Why waste a thread on this?

bhalevy replied:
The thread is required for the boost signal2 callback.
So the callees can .get() while inside the thread.
(The original subscription logic I borrowed from seastar::abort_source didn't need this)

avikivity replied:
There should at least be a comment. Otherwise someone will coroutinize it.

bhalevy commented:
In v9 ([733ca78](https://github.com/scylladb/scylladb/commit/733ca78ff5d46303b1250c9c595f7668083a239d)):
- rebased
- added `_shard0` to disk_space_monitor optional instance in main and cql_est_env
- removed `namespace bs2 = boost::signals2`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13692/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13692/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
:x: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/573/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13251/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5235/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2129/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13692/)
#### Build Details:
- Duration: 4 hr 8 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
In v10 ([9a3db9a](https://github.com/scylladb/scylladb/commit/9a3db9ac9e444e6ef85f52f0223d11cf6d89b4e2)):
- rebased
- focused unit test on space / utilization
  - using loopback mount
- got rid of template for disk_space_monitor
  - now uses only seastar::lowres_clock
- added `_shard0` suffix to disk_space_monitor variable/member in main/cql_test_env

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13829/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13829/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/disk_space_monitor_test
#### Failed Tests (1/1):
- [test_disk_space_monitor_metrics](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13829/testReport/junit/boost/disk_space_monitor_test/test_disk_space_monitor_metrics) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disk_space_monitor_metrics)
#### Build Details:
- Duration: 33 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13829/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13829/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/disk_space_monitor_test
> 
> #### Failed Tests (1/1):
> * [test_disk_space_monitor_metrics](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13829/testReport/junit/boost/disk_space_monitor_test/test_disk_space_monitor_metrics) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disk_space_monitor_metrics)

Bummer, the loop device cannot be used inside podman in rootless mode.

bhalevy commented:
> > #### 🔴 CI State: FAILURE
> > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13829/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13829/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/disk_space_monitor_test
> > #### Failed Tests (1/1):
> > 
> > * [test_disk_space_monitor_metrics](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13829/testReport/junit/boost/disk_space_monitor_test/test_disk_space_monitor_metrics) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disk_space_monitor_metrics)
> 
> Bummer, the loop device cannot be used inside podman in rootless mode.

I will move the test to test/manual so it can run natively on the host when required.

bhalevy commented:
In v11 ([93fd009](https://github.com/scylladb/scylladb/commit/93fd009de2ce9d68a8a58e279faa6bd5612552cd)):
- rebased (resolved merge conflicts)
- moved unit test to test/manual as per https://github.com/scylladb/scylladb/pull/21523#issuecomment-2543832155

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13925/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/766/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13504/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2375/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5492/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/168/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13925/)
#### Build Details:
- Duration: 3 hr 57 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
- removed spurious space on end-of-line.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13957/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/793/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2401/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13531/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5517/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/195/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13957/)
#### Failed Tests (3/36360):
- [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13957/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
- [test_simple_removenode_4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13957/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_4)
- [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13957/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)
#### Build Details:
- Duration: 8 hr 5 min
- Builder: i-06a6ca528f5fbe931 (m5ad.8xlarge)

avikivity commented on test/manual/disk_space_monitor_test.cc:
```diff
@@ -0,0 +1,196 @@
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
```
Outdated license.

avikivity commented on test/manual/disk_space_monitor_test.cc:
```diff
@@ -0,0 +1,196 @@
+    sstring losetup_setup_cmd = fmt::format("sudo losetup --find --show '{}'", image_file.native());
+    auto loop_dev = run_system_command(dir, losetup_setup_cmd).get();
+    auto detach_loop_dev = defer([&] () noexcept {
+        sstring losetup_detach_cmd = fmt::format("sudo losetup --detach '{}'", loop_dev);
```
It's really bad practice to sudo without warning.

bhalevy replied:
What kind of warning do you have in mind?
Just a log message?

avikivity replied:
No, using sudo on behalf of the user without the user knowing beforehand that this is what is going to happen.

Let's drop the test and understand that this is not practically testable.

avikivity replied:
(at least in unit tests)

bhalevy replied:
okay, will drop unit test

avikivity commented on test/manual/disk_space_monitor_test.cc:
```diff
@@ -0,0 +1,196 @@
+
+    BOOST_REQUIRE_GT(sampled_disk_utilization, 0.5);
+    BOOST_REQUIRE_EQUAL(sampled_space.available + st.allocated_size, dsm_space.available);
+}
```
test/manual = don't run me

bhalevy replied:
I started with putting it in test/boost, but it doesn't have root permissions when running under rootless podman.
I can run the test on a live filesystem, like /tmp, but then it would be flaky.
What do you suggest?

avikivity replied:
Hm. Maybe add an option to supply a function to call instead of fsstat.

It would be pretty silly though. What would we be testing?

avikivity commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,87 @@
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
```
Outdated license

bhalevy replied:
Fixed in v12 (5620073786)

avikivity commented on utils/disk_space_monitor.cc:
```diff
@@ -0,0 +1,87 @@
+                get_filesystem_stats().get();
+
+                _signal_source();
+
```
Can you explain how signals2 requires a thread?

bhalevy replied:
`boost::signals2::signal_type` can't return a `future<>`.
But that said, we can call the subscribed callback under a gate (or phased barrier) and wait for it in stop().
I'll do that and get rid of the thread.

bhalevy replied:
Dropped async thread in v12 (5620073786)

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -0,0 +1,84 @@
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
```
Outdated license.

bhalevy replied:
Fixed in v12 (5620073786)

bhalevy commented:
In v12 (5620073786):
- Fixed license for new files
- Used phased_barrier for callbacks
  - For dropping the seastar thread in the poll() loop.

avikivity commented on main.cc:
```diff
@@ -1172,6 +1174,26 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
+            supervisor::notify("starting disk space monitor");
+            auto dsm_cfg = utils::disk_space_monitor::config{
+                .sched_group = dbcfg.compaction_scheduling_group,
```
Why compaction?

Seems to me it's a maintenance task.

bhalevy replied:
When disk utilization is low this could be done in maintenance (streaming) sched group, but when we're low on disk space, should we monitor the filesystem in higher priority?

bhalevy replied:
Hmm, actually, one of the planned use cases is to make streaming shares dynamic, and increase them as we run out of space, we can use the streaming scheduling group all the way.

bhalevy replied:
Fixed in v13 ([723785f](https://github.com/scylladb/scylladb/commit/723785f21bd84fa8895d66851d25fc2b19fd915c))

bhalevy commented:
In v13 ([723785f](https://github.com/scylladb/scylladb/commit/723785f21bd84fa8895d66851d25fc2b19fd915c)):
- rebased
- used streaming_scheduling_group
- dropped manual unit test

bhalevy commented:
Draft until https://github.com/scylladb/seastar/pull/2597 reaches scylladb.git

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14057/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/870/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2500/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13628/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/274/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5615/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14057/)
#### Build Details:
- Duration: 4 hr 3 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
> Draft until [scylladb/seastar#2597](https://github.com/scylladb/seastar/pull/2597) reaches scylladb.git

Part of https://github.com/scylladb/scylladb/pull/22076

bhalevy commented:
v14 ([fce502a](https://github.com/scylladb/scylladb/commit/fce502ac7715258b0be42a893f5c1acff2d1a258)):
- rebased on top of latest seastar submodule update
  - containing scylladb/seastar@a9bef537696b556104d42848629756a8a4b198ca
- use seastar::file_system_space to get the space info

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14348/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13825/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2697/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5812/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/466/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1064/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14348/)
#### Failed Tests (2/36894):
- [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14348/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_snapshot_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
- [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14348/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_snapshot_dev_3_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
#### Build Details:
- Duration: 4 hr 37 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14372/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1084/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/486/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14372/)
> 
> #### Failed Tests (2/29677):
> * [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14372/testReport/junit/topology_custom/test_view_build_status/test_view_build_status_snapshot_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
> * [test_view_build_status_snapshot.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14372/testReport/junit/topology_custom/test_view_build_status/test_view_build_status_snapshot_dev_3_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_snapshot.dev.3)
> 
> #### Build Details:
> * Duration: 8 hr 29 min
> * Builder: i-01406c109897f4ff1 (m5ad.8xlarge)

Unrelated: https://github.com/scylladb/scylladb/issues/20638
